### ISSUE TITLE
[FR] add Gumbel distribution

### ISSUE THREAD

**Issue by fehiepsi** [pyro-ppl/numpyro#548]
As requested @tbsexton at #545, it would be nice to have Gumbel distribution in NumPyro.

References: [wikipedia](https://en.wikipedia.org/wiki/Gumbel_distribution), [pytorch dist](https://pytorch.org/docs/stable/distributions.html#gumbel), [jax sampler](https://jax.readthedocs.io/en/latest/jax.random.html#jax.random.gumbel).

**Comment by tbsexton** [comment]
Also commenting to add info on `gumbel-softmax`, which kicked off this discussion (continuous approximation of categorical distribution). 

[Paper](https://arxiv.org/pdf/1611.01144.pdf), [(deprecated?) Pytorch version with straight-through forward-mode support](https://github.com/pytorch/pytorch/blob/master/torch/nn/functional.py#L1447)

**Comment by tbsexton** [comment]
Wow, thanks to @daydreamt! That was fast. I'd be happy to close this, or re-name it as "Gumbel-Softmax" if that functionality is also wanted. 

Based on the pytorch link above, it seems like they are making use of some pytorch-specific detachments, etc. to implement the straight-through version (argmax in the fwd/sampling direction, softmax in the backward/gradient direction, if I'm understanding correctly). Not sure how such a thing could be done with the JAX backend.

**Comment by fehiepsi** [comment]
@tbsexton A similar version can be implemented using [jax.lax.stop_gradient](https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.stop_gradient.html#jax.lax.stop_gradient), which plays the role of PyTorch `detach`. However, I am not sure how useful it is to solve your problem. Could you open a new issue which sketches your ideas and your feature request? FYI, we have [RelaxedOneHotCategoricalStraightThrough](http://docs.pyro.ai/en/1.2.1/_modules/pyro/distributions/relaxed_straight_through.html#RelaxedOneHotCategoricalStraightThrough) distribution in Pyro, which can be ported to NumPyro without much difficulty.

I would like to close this issue for now and discuss on the new one. :)
